cpp/src/cylon/table.cpp

```diff
@@ -1076,6 +1153,138 @@ Status DistributedEquals(const std::shared_ptr<cylon::Table> &a, const std::shar
   return Status::OK();
 }
 
+std::vector<std::pair<int, int>> find_mapping(int start_idx, int size, std::vector<int64_t> target, std::vector<int64_t> target_acc) {
+  std::vector<std::pair<int, int>> result;
+
+  auto itr_temp = std::upper_bound(target_acc.begin(), target_acc.end(), start_idx);
+  auto itr = std::prev(itr_temp); // last element that less than or equal to start_idx
+
+  int dest_idx = itr - target_acc.begin();
+  int dest_offset = start_idx - target_acc[dest_idx];
+  int ele_left = size;
+
+  while(ele_left > 0) {
+    int cur_dest_req_sz = target[dest_idx] - dest_offset;
+    int send_sz = std::min(ele_left, cur_dest_req_sz);
+    result.push_back({dest_idx, send_sz});
+    if(cur_dest_req_sz <= ele_left) {
+      dest_idx++;
+      dest_offset = 0;
+    }
+    ele_left -= send_sz;
+  }
+
+  return result;
+}
+
+Status Repartition(const std::shared_ptr<cylon::Table>& table, 
+                   const std::vector<int64_t>& rows_per_partition,
+                   const std::vector<int>& receive_build_rank_order,
+                   std::shared_ptr<Table> *output) {
+  int world_size = table->GetContext()->GetWorldSize();
+  int rank = table->GetContext()->GetRank();
+  int num_row = table->Rows();
+  std::vector<int64_t> size = { num_row };
+  std::vector<int64_t> sizes;
+  mpi::AllGather(size, world_size, sizes);
+
+  if(rows_per_partition.size() != world_size) {
+    return Status(
+    cylon::Code::ValueError,
+    "rows_per_partition size does not align with world size. Received " +
+        std::to_string(rows_per_partition.size()) + ", Expected " +
+        std::to_string(world_size));
+  }
+
+  // collecting start index of each process after the repartition
+  std::vector<int64_t> dest_sizes_acc(rows_per_partition.size());
+  dest_sizes_acc[0] = 0;
+  for(int i = 1; i < dest_sizes_acc.size(); i++) {
+    dest_sizes_acc[i] = dest_sizes_acc[i - 1] + rows_per_partition[i - 1];
+  }
+
+  int start_idx = 0;
+  for(int i = 0; i < rank; i++) {
+    start_idx += sizes[i];
+  }
+
+  int acc = start_idx;
+  for(int i = rank; i < world_size; i++) {
+    acc += sizes[i];
+  }
+
+  if(acc != dest_sizes_acc.back() + rows_per_partition.back()) {
+    return Status(
+    cylon::Code::ValueError,
+    "rows_per_partition total number of rows does not align with actual number of rows. Received " +
+        std::to_string(dest_sizes_acc.back() + rows_per_partition.back()) + ", Expected " +
+        std::to_string(acc));
+  }
+
+  std::vector<std::pair<int, int>> send_to = find_mapping(start_idx, num_row, rows_per_partition, dest_sizes_acc);
```

Comment:
I'd rather use a `vector<int64>` instead of `vector<pair<int, int64>>` here. We can make the index of the vector correspond to the rank, isn't it? :thinking: